From 6e9c663f312e9b2e676d544afa4a548970d31249 Mon Sep 17 00:00:00 2001
From: Zheng Song <zheng.song@intel.com>
Date: Mon, 30 May 2022 17:00:04 +0800
Subject: [PATCH] Adapt x86 power control to Puti paltform

Puti platform uses a CPLD to monitor and control power control pins
instead of using BMC to manipulate relevant GPIO directly.
This commit fills the gap. Adapt the x86 power control to the CPLD
interfaces.

Signed-off-by: Zheng Song <zheng.song@intel.com>

---
 config/power-config-host0.json | 106 ++++++----
 src/power_control.cpp          | 372 ++++++++++++++++++++++++++-------
 2 files changed, 352 insertions(+), 126 deletions(-)

diff --git a/config/power-config-host0.json b/config/power-config-host0.json
index 12c3309..411a077 100644
--- a/config/power-config-host0.json
+++ b/config/power-config-host0.json
@@ -2,82 +2,96 @@
   "gpio_configs":[
     {
         "Name" : "IdButton",
-        "LineName" : "ID_BUTTON",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
-    },
-    {
-        "Name" : "NMIButton",
-        "LineName" : "NMI_BUTTON",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Type" : "DBUS_ONE_WAY_SIGNAL",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_events",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformEvents",
+        "Property" : "UidButton"
     },
     {
         "Name" : "NMIOut",
-        "LineName" : "NMI_OUT",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveHigh"
+        "Type" : "DBUS_METHOD",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_controls",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformControls",
+        "Property" : "BmcNmiCtrl"
     },
     {
         "Name" : "PostComplete",
-        "LineName" : "POST_COMPLETE",
+        "LineName" : "FM_BIOS_POST_CMPLT_BMC_N",
         "Type" : "GPIO",
         "Polarity" : "ActiveLow"
     },
     {
-        "Name" : "PowerButton",
-        "LineName" : "POWER_BUTTON",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Name" : "powerButtonShortPressedForOn",
+        "Type" : "DBUS_ONE_WAY_SIGNAL",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_events",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformEvents",
+        "Property" : "ShortBtnPowerOn"
     },
     {
-        "Name" : "PowerOk",
-        "LineName" : "PS_PWROK",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveHigh"
+        "Name" : "powerButtonLongPressedForOff",
+        "Type" : "DBUS_ONE_WAY_SIGNAL",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_events",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformEvents",
+        "Property" : "LongBtnPowerDown"
     },
     {
-        "Name" : "PowerOut",
-        "LineName" : "POWER_OUT",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Name" : "PowerOk",
+        "Type" : "DBUS",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_states",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformStates",
+        "Property" : "PowerOk"
     },
     {
-        "Name" : "ResetButton",
-        "LineName" : "RESET_BUTTON",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Name" : "ShortBtnPowerDownCtrl",
+        "Type" : "DBUS_METHOD",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_controls",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformControls",
+        "Property" : "ShortBtnPowerDownCtrl"
     },
     {
-        "Name" : "ResetOut",
-        "LineName" : "RESET_OUT",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Name" : "LongBtnPowerDownCtrl",
+        "Type" : "DBUS_METHOD",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_controls",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformControls",
+        "Property" : "LongBtnPowerDownCtrl"
     },
     {
-        "Name" : "SioOnControl",
-        "LineName" : "SIO_ONCONTROL",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Name" : "ShortBtnPowerOnCtrl",
+        "Type" : "DBUS_METHOD",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_controls",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformControls",
+        "Property" : "ShortBtnPowerOnCtrl"
     },
     {
-        "Name" : "SioPowerGood",
-        "LineName" : "SIO_POWER_GOOD",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveHigh"
+        "Name" : "ResetButton",
+        "Type" : "DBUS_ONE_WAY_SIGNAL",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_events",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformEvents",
+        "Property" : "ShortBtnReset"
     },
     {
-        "Name" : "SIOS5",
-        "LineName" : "SIO_S5",
-        "Type" : "GPIO",
-        "Polarity" : "ActiveLow"
+        "Name" : "ResetCtrl",
+        "Type" : "DBUS_METHOD",
+        "DbusName" : "xyz.openbmc_project.Cpld",
+        "Path" : "/xyz/openbmc_project/cpld/platform_controls",
+        "Interface" : "xyz.openbmc_project.Cpld.PlatformControls",
+        "Property" : "ResetCtrl"
     }
 ],
   "timing_configs":{
       "PowerPulseMs": 200,
       "ForceOffPulseMs": 15000,
       "ResetPulseMs": 500,
-      "PowerCycleMs": 5000,
+      "PowerCycleMs": 8000,
       "SioPowerGoodWatchdogMs": 12000,
       "PsPowerOKWatchdogMs": 8000,
       "GracefulPowerOffS": 300,
diff --git a/src/power_control.cpp b/src/power_control.cpp
index 21c19dd..aaf4607 100644
--- a/src/power_control.cpp
+++ b/src/power_control.cpp
@@ -53,7 +53,9 @@ boost::container::flat_map<DbusConfigType, std::string> dbusParams = {
 enum class ConfigType
 {
     GPIO = 1,
-    DBUS
+    DBUS,                   // includes 2 direction signals
+    DBUS_ONE_WAY_SIGNAL,    // only assert signal
+    DBUS_METHOD             // for outputs only
 };
 
 struct ConfigData
@@ -68,14 +70,20 @@ struct ConfigData
 };
 
 static ConfigData powerOutConfig;
+static ConfigData shortBtnPowerDownCtrlConfig;
+static ConfigData longBtnPowerDownCtrlConfig;
+static ConfigData shortBtnPowerOnCtrlConfig;
 static ConfigData powerOkConfig;
 static ConfigData resetOutConfig;
+static ConfigData resetCtrlConfig;
 static ConfigData nmiOutConfig;
 static ConfigData sioPwrGoodConfig;
 static ConfigData sioOnControlConfig;
 static ConfigData sioS5Config;
 static ConfigData postCompleteConfig;
 static ConfigData powerButtonConfig;
+static ConfigData powerButtonShortPressedForOnConfig;
+static ConfigData powerButtonLongPressedForOffConfig;
 static ConfigData resetButtonConfig;
 static ConfigData idButtonConfig;
 static ConfigData nmiButtonConfig;
@@ -85,14 +93,20 @@ static ConfigData slotPowerConfig;
 // power control json config
 boost::container::flat_map<std::string, ConfigData*> powerSignalMap = {
     {"PowerOut", &powerOutConfig},
+    {"ShortBtnPowerDownCtrl", &shortBtnPowerDownCtrlConfig},
+    {"LongBtnPowerDownCtrl", &longBtnPowerDownCtrlConfig},
+    {"ShortBtnPowerOnCtrl", &shortBtnPowerOnCtrlConfig},
     {"PowerOk", &powerOkConfig},
     {"ResetOut", &resetOutConfig},
+    {"ResetCtrl", &resetCtrlConfig},
     {"NMIOut", &nmiOutConfig},
     {"SioPowerGood", &sioPwrGoodConfig},
     {"SioOnControl", &sioOnControlConfig},
     {"SIOS5", &sioS5Config},
     {"PostComplete", &postCompleteConfig},
     {"PowerButton", &powerButtonConfig},
+    {"powerButtonShortPressedForOn", &powerButtonShortPressedForOnConfig},
+    {"powerButtonLongPressedForOff", &powerButtonLongPressedForOffConfig},
     {"ResetButton", &resetButtonConfig},
     {"IdButton", &idButtonConfig},
     {"NMIButton", &nmiButtonConfig},
@@ -277,6 +291,8 @@ enum class Event
     postCompleteAssert,
     postCompleteDeAssert,
     powerButtonPressed,
+    powerButtonShortPressedForOn,
+    powerButtonLongPressedForOff,
     resetButtonPressed,
     powerCycleTimerExpired,
     psPowerOKWatchdogTimerExpired,
@@ -327,6 +343,12 @@ static std::string getEventName(Event event)
         case Event::powerButtonPressed:
             return "power button pressed";
             break;
+        case Event::powerButtonShortPressedForOn:
+            return "power button short pressed for power on";
+            break;
+        case Event::powerButtonLongPressedForOff:
+            return "power button long pressed for power off";
+            break;
         case Event::resetButtonPressed:
             return "reset button pressed";
             break;
@@ -1200,8 +1222,35 @@ static int assertGPIOForMs(const ConfigData& config, const int durationMs)
     return setGPIOOutputForMs(config, config.polarity, durationMs);
 }
 
+static void dbusGPIOMethodCall(ConfigData cfg)
+{
+    if (cfg.type == ConfigType::DBUS_METHOD)
+    {
+        conn->async_method_call(
+            [&](boost::system::error_code ec) {
+                if (ec)
+                {
+                    lg2::info("failed to do {LINE_NAME}",
+                              "LINE_NAME", cfg.lineName);
+                }
+            },
+            cfg.dbusName, cfg.path,
+            cfg.interface, cfg.lineName);
+    }
+    else
+    {
+        lg2::error("error calling to dbusGPIOMethodCall");
+    }
+}
+
 static void powerOn()
 {
+    if (!shortBtnPowerOnCtrlConfig.lineName.empty())
+    {
+        dbusGPIOMethodCall(shortBtnPowerOnCtrlConfig);
+        return;
+    }
+
     assertGPIOForMs(powerOutConfig, TimerMap["PowerPulseMs"]);
 }
 #ifdef CHASSIS_SYSTEM_RESET
@@ -1279,11 +1328,22 @@ static void slotPowerCycle()
 #endif
 static void gracefulPowerOff()
 {
+    if (!shortBtnPowerDownCtrlConfig.lineName.empty())
+    {
+        dbusGPIOMethodCall(shortBtnPowerDownCtrlConfig);
+        return;
+    }
     assertGPIOForMs(powerOutConfig, TimerMap["PowerPulseMs"]);
 }
 
 static void forcePowerOff()
 {
+    if (!longBtnPowerDownCtrlConfig.lineName.empty())
+    {
+        dbusGPIOMethodCall(longBtnPowerDownCtrlConfig);
+        return;
+    }
+
     if (assertGPIOForMs(powerOutConfig, TimerMap["ForceOffPulseMs"]) < 0)
     {
         return;
@@ -1309,6 +1369,11 @@ static void forcePowerOff()
 
 static void reset()
 {
+    if (!resetCtrlConfig.lineName.empty())
+    {
+        dbusGPIOMethodCall(resetCtrlConfig);
+        return;
+    }
     assertGPIOForMs(resetOutConfig, TimerMap["ResetPulseMs"]);
 }
 
@@ -1607,6 +1672,9 @@ static void powerStateOn(const Event event)
             setPowerState(PowerState::gracefulTransitionToOff);
             gracefulPowerOffTimerStart();
             break;
+        case Event::powerButtonLongPressedForOff:
+            setPowerState(PowerState::transitionToOff);
+            break;
         case Event::powerOffRequest:
             setPowerState(PowerState::transitionToOff);
             forcePowerOff();
@@ -1709,6 +1777,7 @@ static void powerStateOff(const Event event)
         case Event::sioS5DeAssert:
             setPowerState(PowerState::waitForPSPowerOK);
             break;
+        case Event::powerButtonShortPressedForOn:
         case Event::powerButtonPressed:
             psPowerOKWatchdogTimerStart();
             setPowerState(PowerState::waitForPSPowerOK);
@@ -1752,6 +1821,9 @@ static void powerStateGracefulTransitionToOff(const Event event)
         case Event::gracefulPowerOffTimerExpired:
             setPowerState(PowerState::on);
             break;
+        case Event::powerButtonLongPressedForOff:
+            setPowerState(PowerState::transitionToOff);
+            break;
         case Event::powerOffRequest:
             gracefulPowerOffTimer.cancel();
             setPowerState(PowerState::transitionToOff);
@@ -1796,6 +1868,7 @@ static void powerStateCycleOff(const Event event)
             powerCycleTimer.cancel();
             setPowerState(PowerState::waitForPSPowerOK);
             break;
+        case Event::powerButtonShortPressedForOn:
         case Event::powerButtonPressed:
             powerCycleTimer.cancel();
             psPowerOKWatchdogTimerStart();
@@ -1913,6 +1986,36 @@ static void sioS5Handler(bool state)
     sendPowerControlEvent(powerControlEvent);
 }
 
+static void powerButtonShortPressedForOnHandler(bool state)
+{
+    powerButtonIface->set_property("ButtonPressed", true);
+    powerButtonIface->set_property("ButtonPressed", false);
+    lg2::info("powerButtonShortPressedOnHandler");
+    powerButtonPressLog();
+    sendPowerControlEvent(Event::powerButtonShortPressedForOn);
+    addRestartCause(RestartCause::powerButton);
+}
+
+static void powerButtonLongPressedForOffHandler(bool state)
+{
+    powerButtonIface->set_property("ButtonPressed", true);
+    powerButtonIface->set_property("ButtonPressed", false);
+    lg2::info("powerButtonLongPressedForOffHandler");
+    powerButtonPressLog();
+    sendPowerControlEvent(Event::powerButtonLongPressedForOff);
+}
+
+static void powerButtonShortPressedForResetHandler(bool state)
+{
+    resetButtonIface->set_property("ButtonPressed", true);
+    resetButtonIface->set_property("ButtonPressed", false);
+    lg2::info("powerButtonShortPressedForResetHandler");
+    resetButtonPressLog();
+    sendPowerControlEvent(Event::resetButtonPressed);
+    addRestartCause(RestartCause::resetButton);
+}
+
+
 static void powerButtonHandler(bool state)
 {
     powerButtonIface->set_property("ButtonPressed", !state);
@@ -1933,18 +2036,29 @@ static void powerButtonHandler(bool state)
 
 static void resetButtonHandler(bool state)
 {
-    resetButtonIface->set_property("ButtonPressed", !state);
-    if (!state)
+    if (resetButtonConfig.type == ConfigType::DBUS_ONE_WAY_SIGNAL)
     {
+        resetButtonIface->set_property("ButtonPressed", true);
+        resetButtonIface->set_property("ButtonPressed", false);
         resetButtonPressLog();
-        if (!resetButtonMask)
-        {
-            sendPowerControlEvent(Event::resetButtonPressed);
-            addRestartCause(RestartCause::resetButton);
-        }
-        else
+        sendPowerControlEvent(Event::resetButtonPressed);
+        addRestartCause(RestartCause::resetButton);
+    }
+    else
+    {
+        resetButtonIface->set_property("ButtonPressed", !state);
+        if (!state)
         {
-            lg2::info("reset button press masked");
+            resetButtonPressLog();
+            if (!resetButtonMask)
+            {
+                sendPowerControlEvent(Event::resetButtonPressed);
+                addRestartCause(RestartCause::resetButton);
+            }
+            else
+            {
+                lg2::info("reset button press masked");
+            }
         }
     }
 }
@@ -1988,30 +2102,38 @@ static void nmiSetEnableProperty(bool value)
 
 static void nmiReset(void)
 {
-    static constexpr const uint8_t value = 1;
-    const static constexpr int nmiOutPulseTimeMs = 200;
+    if (nmiOutConfig.type == ConfigType::GPIO)
+    {
+        static constexpr const uint8_t value = 1;
+        const static constexpr int nmiOutPulseTimeMs = 200;
 
-    lg2::info("NMI out action");
-    nmiOutLine.set_value(value);
-    lg2::info("{GPIO_NAME} set to {GPIO_VALUE}", "GPIO_NAME",
-              nmiOutConfig.lineName, "GPIO_VALUE", value);
-    gpioAssertTimer.expires_after(std::chrono::milliseconds(nmiOutPulseTimeMs));
-    gpioAssertTimer.async_wait([](const boost::system::error_code ec) {
-        // restore the NMI_OUT GPIO line back to the opposite value
-        nmiOutLine.set_value(!value);
-        lg2::info("{GPIO_NAME} released", "GPIO_NAME", nmiOutConfig.lineName);
-        if (ec)
-        {
-            // operation_aborted is expected if timer is canceled before
-            // completion.
-            if (ec != boost::asio::error::operation_aborted)
+        lg2::info("NMI out action");
+        nmiOutLine.set_value(value);
+        lg2::info("{GPIO_NAME} set to {GPIO_VALUE}", "GPIO_NAME",
+                nmiOutConfig.lineName, "GPIO_VALUE", value);
+        gpioAssertTimer.expires_after(std::chrono::milliseconds(nmiOutPulseTimeMs));
+        gpioAssertTimer.async_wait([](const boost::system::error_code ec) {
+            // restore the NMI_OUT GPIO line back to the opposite value
+            nmiOutLine.set_value(!value);
+            lg2::info("{GPIO_NAME} released", "GPIO_NAME", nmiOutConfig.lineName);
+            if (ec)
             {
-                lg2::error("{GPIO_NAME} async_wait failed: {ERROR_MSG}",
-                           "GPIO_NAME", nmiOutConfig.lineName, "ERROR_MSG",
-                           ec.message());
+                // operation_aborted is expected if timer is canceled before
+                // completion.
+                if (ec != boost::asio::error::operation_aborted)
+                {
+                    lg2::error("{GPIO_NAME} async_wait failed: {ERROR_MSG}",
+                            "GPIO_NAME", nmiOutConfig.lineName, "ERROR_MSG",
+                            ec.message());
+                }
             }
-        }
-    });
+        });
+    }
+    else if (nmiOutConfig.type == ConfigType::DBUS_METHOD)
+    {
+        dbusGPIOMethodCall(nmiOutConfig);
+    }
+
     // log to redfish
     nmiDiagIntLog();
     lg2::info("NMI out action completed");
@@ -2100,7 +2222,13 @@ static void nmiButtonHandler(bool state)
 
 static void idButtonHandler(bool state)
 {
-    idButtonIface->set_property("ButtonPressed", !state);
+    if (idButtonConfig.type == ConfigType::DBUS_ONE_WAY_SIGNAL)
+    {
+        idButtonIface->set_property("ButtonPressed", true);
+        idButtonIface->set_property("ButtonPressed", false);
+    }
+    else
+        idButtonIface->set_property("ButtonPressed", !state);
 }
 
 static void pltRstHandler(bool pltRst)
@@ -2229,6 +2357,14 @@ static int loadConfigValues()
         {
             tempGpioData->type = ConfigType::DBUS;
         }
+        else if (signalType == "DBUS_ONE_WAY_SIGNAL")
+        {
+            tempGpioData->type = ConfigType::DBUS_ONE_WAY_SIGNAL;
+        }
+        else if (signalType == "DBUS_METHOD")
+        {
+            tempGpioData->type = ConfigType::DBUS_METHOD;
+        }
         else
         {
             lg2::error("{TYPE} is not a recognized power-control signal type",
@@ -2364,7 +2500,23 @@ static sdbusplus::bus::match::match
         std::move(pulseEventMatcherCallback));
 }
 
-int getProperty(ConfigData& configData)
+static sdbusplus::bus::match::match
+    dbusOneWaySignalMatcher(const ConfigData& cfg,
+                            std::function<void(bool)> onMatch)
+{
+    auto pulseEventMatcherCallback =
+        [&cfg, onMatch](sdbusplus::message::message& msg) {
+            onMatch(false);
+        };
+
+    return sdbusplus::bus::match::match(
+        static_cast<sdbusplus::bus::bus&>(*conn),
+        "type='signal',interface='" + cfg.interface + "'," +
+        "member='" + cfg.lineName + "'",
+        std::move(pulseEventMatcherCallback));
+}
+
+bool getProperty(ConfigData& configData)
 {
     auto method = conn->new_method_call(
         configData.dbusName.c_str(), configData.path.c_str(),
@@ -2378,7 +2530,7 @@ int getProperty(ConfigData& configData)
             "Error reading {PROPERTY} D-Bus property on interface {INTERFACE} and path {PATH}",
             "PROPERTY", configData.lineName, "INTERFACE", configData.interface,
             "PATH", configData.path);
-        return -1;
+        return false;
     }
     std::variant<bool> resp;
     reply.read(resp);
@@ -2387,7 +2539,7 @@ int getProperty(ConfigData& configData)
     {
         lg2::error("Error: {PROPERTY} D-Bus property is not the expected type",
                    "PROPERTY", configData.lineName);
-        return -1;
+        return false;
     }
     return (*respValue);
 }
@@ -2540,22 +2692,40 @@ int main(int argc, char* argv[])
         }
     }
 
-    // Request POWER_BUTTON GPIO events
-    if (powerButtonConfig.type == ConfigType::GPIO)
+    if (!powerButtonShortPressedForOnConfig.lineName.empty())
     {
-        if (!requestGPIOEvents(powerButtonConfig.lineName, powerButtonHandler,
-                               powerButtonLine, powerButtonEvent))
-        {
+        if (powerButtonLongPressedForOffConfig.lineName.empty())
             return -1;
-        }
+
+        if (powerButtonShortPressedForOnConfig.type != ConfigType::DBUS_ONE_WAY_SIGNAL
+            || powerButtonLongPressedForOffConfig.type != ConfigType::DBUS_ONE_WAY_SIGNAL)
+            return -1;
+        static sdbusplus::bus::match::match PBShortPressedForOntMonitor =
+                power_control::dbusOneWaySignalMatcher(powerButtonShortPressedForOnConfig,
+                                            powerButtonShortPressedForOnHandler);
+        static sdbusplus::bus::match::match PBLongPressedForOfftMonitor =
+                power_control::dbusOneWaySignalMatcher(powerButtonLongPressedForOffConfig,
+                                            powerButtonLongPressedForOffHandler);
     }
-    else if (powerButtonConfig.type == ConfigType::DBUS)
+    else
     {
-        static sdbusplus::bus::match::match powerButtonEventMonitor =
-            power_control::dbusGPIOMatcher(powerButtonConfig,
-                                           powerButtonHandler);
+        // Request POWER_BUTTON GPIO events
+        if (powerButtonConfig.type == ConfigType::GPIO)
+        {
+            if (!requestGPIOEvents(powerButtonConfig.lineName, powerButtonHandler,
+                                powerButtonLine, powerButtonEvent))
+            {
+                return -1;
+            }
+        }
+        else if (powerButtonConfig.type == ConfigType::DBUS)
+        {
+            static sdbusplus::bus::match::match powerButtonEventMonitor =
+                power_control::dbusGPIOMatcher(powerButtonConfig,
+                                            powerButtonHandler);
+        }
     }
-
+    
     // Request RESET_BUTTON GPIO events
     if (resetButtonConfig.type == ConfigType::GPIO)
     {
@@ -2571,6 +2741,12 @@ int main(int argc, char* argv[])
             power_control::dbusGPIOMatcher(resetButtonConfig,
                                            resetButtonHandler);
     }
+    else if (resetButtonConfig.type == ConfigType::DBUS_ONE_WAY_SIGNAL)
+    {
+        static sdbusplus::bus::match::match resetButtonEventMonitor =
+            power_control::dbusOneWaySignalMatcher(resetButtonConfig,
+                                           resetButtonHandler);
+    }
 
     // Request NMI_BUTTON GPIO events
     if (nmiButtonConfig.type == ConfigType::GPIO)
@@ -2601,6 +2777,12 @@ int main(int argc, char* argv[])
         static sdbusplus::bus::match::match idButtonEventMonitor =
             power_control::dbusGPIOMatcher(idButtonConfig, idButtonHandler);
     }
+    else if (idButtonConfig.type == ConfigType::DBUS_ONE_WAY_SIGNAL)
+    {
+        static sdbusplus::bus::match::match idButtonEventMonitor =
+            power_control::dbusOneWaySignalMatcher(idButtonConfig,
+                                           idButtonHandler);
+    }
 
 #ifdef USE_PLT_RST
     sdbusplus::bus::match::match pltRstMatch(
@@ -2633,42 +2815,52 @@ int main(int argc, char* argv[])
     }
 
     // initialize NMI_OUT GPIO.
-    if (!nmiOutConfig.lineName.empty())
+    if (!nmiOutConfig.lineName.empty()
+        && nmiOutConfig.type != ConfigType::DBUS_METHOD)
     {
         setGPIOOutput(nmiOutConfig.lineName, 0, nmiOutLine);
     }
 
     // Initialize POWER_OUT and RESET_OUT GPIO.
     gpiod::line line;
-    if (!powerOutConfig.lineName.empty())
+    if (shortBtnPowerDownCtrlConfig.lineName.empty())
     {
-        if (!setGPIOOutput(powerOutConfig.lineName, !powerOutConfig.polarity,
-                           line))
+        if (!powerOutConfig.lineName.empty())
+        {
+            if (!setGPIOOutput(
+                    powerOutConfig.lineName, !powerOutConfig.polarity, line))
+            {
+                return -1;
+            }
+        }
+        else
         {
+            lg2::error(
+                "powerOut name should be configured from json config file");
             return -1;
         }
     }
-    else
-    {
-        lg2::error("powerOut name should be configured from json config file");
-        return -1;
-    }
 
-    if (!resetOutConfig.lineName.empty())
+    if (resetCtrlConfig.lineName.empty())
     {
-        if (!setGPIOOutput(resetOutConfig.lineName, !resetOutConfig.polarity,
-                           line))
+        if (!resetOutConfig.lineName.empty())
         {
+            if (!setGPIOOutput(
+                    resetOutConfig.lineName, !resetOutConfig.polarity, line))
+            {
+                return -1;
+            }
+        }
+        else
+        {
+            lg2::error(
+                "ResetOut name should be configured from json config file");
             return -1;
         }
     }
-    else
-    {
-        lg2::error("ResetOut name should be configured from json config file");
-        return -1;
-    }
     // Release line
-    line.reset();
+    if (line)
+        line.reset();
 
     // Initialize the power state
     powerState = PowerState::off;
@@ -2688,6 +2880,7 @@ int main(int argc, char* argv[])
             powerState = PowerState::on;
         }
     }
+
     // Initialize the power state storage
     if (initializePowerStateStorage() < 0)
     {
@@ -2697,7 +2890,7 @@ int main(int argc, char* argv[])
     // Check if we need to start the Power Restore policy
     powerRestorePolicyCheck();
 
-    if (nmiOutLine)
+    if (!nmiOutConfig.lineName.empty())
         nmiSourcePropertyMonitor();
 
     lg2::info("Initializing power state.");
@@ -2981,7 +3174,8 @@ int main(int argc, char* argv[])
     sdbusplus::asio::object_server buttonsServer =
         sdbusplus::asio::object_server(conn);
 
-    if (!powerButtonConfig.lineName.empty())
+    if (!powerButtonConfig.lineName.empty()
+        || !powerButtonShortPressedForOnConfig.lineName.empty())
     {
         // Power Button Interface
         power_control::powerButtonIface = buttonsServer.add_interface(
@@ -2996,12 +3190,15 @@ int main(int argc, char* argv[])
                     {
                         return 1;
                     }
-                    if (!setGPIOOutput(powerOutConfig.lineName,
+                    if (shortBtnPowerDownCtrlConfig.lineName.empty())
+                    {
+                        if (!setGPIOOutput(powerOutConfig.lineName,
                                        !powerOutConfig.polarity,
                                        powerButtonMask))
-                    {
-                        throw std::runtime_error("Failed to request GPIO");
-                        return 0;
+                        {
+                            throw std::runtime_error("Failed to request GPIO");
+                            return 0;
+                        }
                     }
                     lg2::info("Power Button Masked.");
                 }
@@ -3021,7 +3218,11 @@ int main(int argc, char* argv[])
 
         // Check power button state
         bool powerButtonPressed;
-        if (powerButtonConfig.type == ConfigType::GPIO)
+        if (!powerButtonShortPressedForOnConfig.lineName.empty())
+        {
+            powerButtonPressed = false;
+        }
+        else if (powerButtonConfig.type == ConfigType::GPIO)
         {
             powerButtonPressed = powerButtonLine.get_value() == 0;
         }
@@ -3052,12 +3253,15 @@ int main(int argc, char* argv[])
                     {
                         return 1;
                     }
-                    if (!setGPIOOutput(resetOutConfig.lineName,
+                    if (resetCtrlConfig.lineName.empty())
+                    {
+                        if (!setGPIOOutput(resetOutConfig.lineName,
                                        !resetOutConfig.polarity,
                                        resetButtonMask))
-                    {
-                        throw std::runtime_error("Failed to request GPIO");
-                        return 0;
+                        {
+                            throw std::runtime_error("Failed to request GPIO");
+                            return 0;
+                        }
                     }
                     lg2::info("Reset Button Masked.");
                 }
@@ -3081,10 +3285,14 @@ int main(int argc, char* argv[])
         {
             resetButtonPressed = resetButtonLine.get_value() == 0;
         }
-        else
+        else if (resetButtonConfig.type == ConfigType::DBUS)
         {
             resetButtonPressed = getProperty(resetButtonConfig) == 0;
         }
+        else
+        {
+            resetButtonPressed = false;
+        }
 
         resetButtonIface->register_property("ButtonPressed",
                                             resetButtonPressed);
@@ -3092,7 +3300,7 @@ int main(int argc, char* argv[])
         resetButtonIface->initialize();
     }
 
-    if (nmiButtonLine)
+    if (!nmiButtonConfig.lineName.empty())
     {
         // NMI Button Interface
         nmiButtonIface = buttonsServer.add_interface(
@@ -3137,7 +3345,7 @@ int main(int argc, char* argv[])
         nmiButtonIface->initialize();
     }
 
-    if (nmiOutLine)
+    if (!nmiOutConfig.lineName.empty())
     {
         // NMI out Service
         sdbusplus::asio::object_server nmiOutServer =
@@ -3151,7 +3359,7 @@ int main(int argc, char* argv[])
         nmiOutIface->initialize();
     }
 
-    if (idButtonLine)
+    if (!idButtonConfig.lineName.empty())
     {
         // ID Button Interface
         idButtonIface = buttonsServer.add_interface(
@@ -3164,10 +3372,14 @@ int main(int argc, char* argv[])
         {
             idButtonPressed = idButtonLine.get_value() == 0;
         }
-        else
+        else if (idButtonConfig.type == ConfigType::DBUS)
         {
             idButtonPressed = getProperty(idButtonConfig) == 0;
         }
+        else
+        {
+            idButtonPressed = false;
+        }
 
         idButtonIface->register_property("ButtonPressed", idButtonPressed);
 
